python/tchannel/tornado/peer.py

```diff
@@ -525,160 +525,123 @@ def send(self, arg1, arg2, arg3,
-            request,
-        )
+        if self._hostport:
+            attempt_times = 1
 
-        try:
-            response = yield self.send_with_retry(
-                request, peer, attempt_times, retry_delay
-            )
-        except ProtocolError as protocol_error:
-            # event: after_receive_protocol_error
-            self.peer.tchannel.event_emitter.fire(
-                EventType.after_receive_system_error,
-                request,
-                protocol_error,
-            )
-            raise
-        except TimeoutError as operational_error:
-            # event: on_operational_error
-            self.peer.tchannel.event_emitter.fire(
-                EventType.on_operational_error,
-                request,
-                operational_error,
-            )
-            raise
-
-        log.debug("Got response %s", response)
-        # event: after_receive_response
-        self.peer.tchannel.event_emitter.fire(
-            EventType.after_receive_response,
-            request,
-            response,
+        response = yield self.send_with_retry(
+            endpoint, arg2, arg3, headers, trace_id,
+            parent_span_id, traceflag, ttl, attempt_times, retry_delay,
         )
 
         raise gen.Return(response)
 
     @gen.coroutine
-    def _send(self, connection, req):
-        response_future = connection.send_request(req)
-        with timeout(response_future, req.ttl):
-            response = yield response_future
-
-        raise gen.Return(response)
-
-    @gen.coroutine
-    def send_with_retry(self, request, peer, attempt_times, retry_delay):
+    def send_with_retry(self, endpoint, arg2, arg3, headers,
+                        trace_id, parent_span_id, traceflag,
+                        ttl, attempt_times, retry_delay):
         # black list to record all used peers, so they aren't chosen again.
         blacklist = set()
         response = None
-        connection = yield peer.connect()
-        # mac number of times to retry 3.
+        peer = self.peer
+        # max number of times to retry 3.
         for num_of_attempt in range(attempt_times):
             try:
-                response = yield self._send(connection, request)
-                break
-            except ProtocolError as protocol_error:
-                # event: after_receive_protocol_error_per_attempt
-                self.peer.tchannel.event_emitter.fire(
-                    EventType.after_receive_system_error_per_attempt,
-                    request,
-                    protocol_error,
+                connection = yield peer.connect()
+                request = Request(
+                    service=self.service,
+                    argstreams=[
+                        InMemStream(endpoint), arg2.clone(), arg3.clone()
+                    ],
+                    id=connection.next_message_id(),
+                    headers=headers,
+                    endpoint=endpoint,
```

Comment:
For a different CR: Why is the endpoint being duplicated here? `argstreams` contains it too. We should probably change this since we decided that endpoints cannot be streamed.